File: android/src/main/kotlin/com/margelo/nitro/imagespdf/ExifPdfTransform.kt
package com.margelo.nitro.imagespdf

import android.graphics.Matrix
import android.graphics.RectF
import androidx.exifinterface.media.ExifInterface

internal data class PdfMatrix(
  val a: Float,
  val b: Float,
  val c: Float,
  val d: Float,
  val e: Float,
  val f: Float
)

internal data class ExifPdfTransform(
  val rotationDegrees: Int,
  val isFlipped: Boolean
) {
  val swapsDimensions: Boolean
    get() = rotationDegrees == 90 || rotationDegrees == 270

  fun bitmapMatrix(rawWidth: Int, rawHeight: Int): Matrix {
    val matrix = Matrix()
    if (isFlipped) {
      matrix.postScale(-1f, 1f)
    }
    if (rotationDegrees != 0) {
      matrix.postRotate(rotationDegrees.toFloat())
    }

    val bounds = RectF(0f, 0f, rawWidth.toFloat(), rawHeight.toFloat())
    matrix.mapRect(bounds)
    matrix.postTranslate(-bounds.left, -bounds.top)
    return matrix
  }

  fun pdfMatrix(
    pixelWidth: Int,
    pixelHeight: Int,
    destination: RectF
  ): PdfMatrix {
    val orientation = bitmapMatrix(pixelWidth, pixelHeight)
    val values = FloatArray(9)
    orientation.getValues(values)
    val orientedWidth = if (swapsDimensions) pixelHeight else pixelWidth
    val orientedHeight = if (swapsDimensions) pixelWidth else pixelHeight
    val scaleX = destination.width() / orientedWidth.toFloat()
    val scaleY = destination.height() / orientedHeight.toFloat()
    val a = values[Matrix.MSCALE_X]
    val b = values[Matrix.MSKEW_Y]
    val c = values[Matrix.MSKEW_X]
    val d = values[Matrix.MSCALE_Y]
    val e = values[Matrix.MTRANS_X]
    val f = values[Matrix.MTRANS_Y]

    // PDF image space has a top-left origin; PDF page space has a bottom-left origin.
    return PdfMatrix(
      // PDF image XObjects are drawn in a normalized unit square. Convert the
      // pixel-space orientation matrix into that coordinate system before
      // applying the destination rectangle.
      a = scaleX * a * pixelWidth,
      b = -scaleY * b * pixelWidth,
      c = -scaleX * c * pixelHeight,
      d = scaleY * d * pixelHeight,
      e = scaleX * (e + c * pixelHeight) + destination.left,
      f = destination.top + destination.height() - scaleY * (f + d * pixelHeight)
    )
  }

  companion object {
    val IDENTITY = ExifPdfTransform(rotationDegrees = 0, isFlipped = false)

    fun from(exif: ExifInterface): ExifPdfTransform {
      return ExifPdfTransform(
        rotationDegrees = exif.rotationDegrees,
        isFlipped = exif.isFlipped
      )
    }
  }
}

File: ios/ExifPdfTransform.swift
import CoreGraphics

struct PdfMatrix {
  let a: CGFloat
  let b: CGFloat
  let c: CGFloat
  let d: CGFloat
  let e: CGFloat
  let f: CGFloat
}

struct ExifPdfTransform {
  static let identity = ExifPdfTransform(orientation: 1)

  let orientation: Int

  init(orientation: Int) {
    self.orientation = (1...8).contains(orientation) ? orientation : 1
  }

  var swapsDimensions: Bool {
    orientation >= 5 && orientation <= 8
  }

  func pdfMatrix(
    pixelWidth: Int,
    pixelHeight: Int,
    destination: CGRect
  ) -> PdfMatrix {
    let orientedWidth = swapsDimensions ? pixelHeight : pixelWidth
    let orientedHeight = swapsDimensions ? pixelWidth : pixelHeight
    let scaleX = destination.width / CGFloat(orientedWidth)
    let scaleY = destination.height / CGFloat(orientedHeight)
    let transform: PdfMatrix
    switch orientation {
      case 2:
        transform = PdfMatrix(a: -1, b: 0, c: 0, d: 1, e: CGFloat(pixelWidth), f: 0)
      case 3:
        transform = PdfMatrix(
          a: -1, b: 0, c: 0, d: -1,
          e: CGFloat(pixelWidth), f: CGFloat(pixelHeight)
        )
      case 4:
        transform = PdfMatrix(a: 1, b: 0, c: 0, d: -1, e: 0, f: CGFloat(pixelHeight))
      case 5:
        transform = PdfMatrix(a: 0, b: 1, c: 1, d: 0, e: 0, f: 0)
      case 6:
        transform = PdfMatrix(a: 0, b: 1, c: -1, d: 0, e: CGFloat(pixelHeight), f: 0)
      case 7:
        transform = PdfMatrix(
          a: 0, b: -1, c: -1, d: 0,
          e: CGFloat(pixelHeight), f: CGFloat(pixelWidth)
        )
      case 8:
        transform = PdfMatrix(a: 0, b: -1, c: 1, d: 0, e: 0, f: CGFloat(pixelWidth))
      default:
        transform = PdfMatrix(a: 1, b: 0, c: 0, d: 1, e: 0, f: 0)
    }

    // PDF image space has a top-left origin; PDF page space has a bottom-left origin.
    return PdfMatrix(
      // PDF image XObjects are drawn in a normalized unit square. Convert the
      // pixel-space orientation matrix into that coordinate system before
      // applying the destination rectangle.
      a: scaleX * transform.a * CGFloat(pixelWidth),
      b: -scaleY * transform.b * CGFloat(pixelWidth),
      c: -scaleX * transform.c * CGFloat(pixelHeight),
      d: scaleY * transform.d * CGFloat(pixelHeight),
      e: scaleX * (transform.e + transform.c * CGFloat(pixelHeight)) + destination.minX,
      f: destination.minY + destination.height
        - scaleY * (transform.f + transform.d * CGFloat(pixelHeight))
    )
  }
}
